tests/components/demo/test_notify.py

```diff
@@ -8,189 +9,198 @@
-        assert self.hass.services.has_service("notify", "demo") is not None
-        service = "demo_test_target_name"
-        assert self.hass.services.has_service("notify", service) is not None
-
-    def test_messages_to_targets_route(self):
-        """Test message routing to specific target services."""
-        self._setup_notify()
-        self.hass.bus.listen_once("notify", self.record_calls)
-
-        self.hass.services.call(
-            "notify",
-            "demo_test_target_name",
-            {"message": "my message", "title": "my title", "data": {"hello": "world"}},
-        )
 
-        self.hass.block_till_done()
+@pytest.fixture
+def calls():
+    """Fixture to calls."""
+    return []
+
 
-        data = self.calls[0][0].data
+@pytest.fixture
+def record_calls(calls):
+    """Fixture to record calls."""
 
-        assert {
-            "message": "my message",
-            "target": ["test target id"],
-            "title": "my title",
-            "data": {"hello": "world"},
-        } == data
+    @callback
+    def record_calls(*args):
+        """Record calls."""
+        calls.append(args)
+
+    return record_calls
+
+
+@pytest.fixture(name="mock_demo_notify")
+def mock_demo_notify_fixture():
+    """Mock demo notify service."""
+    with patch("homeassistant.components.demo.notify.get_service", autospec=True) as ns:
+        yield ns
+
+
+async def setup_notify(hass):
+    """Test setup."""
+    with assert_setup_component(1, notify.DOMAIN) as config:
+        assert await async_setup_component(hass, notify.DOMAIN, CONFIG)
+    assert config[notify.DOMAIN]
+    await hass.async_block_till_done()
+
+
+async def test_no_notify_service(hass, mock_demo_notify, caplog):
+    """Test missing platform notify service instance."""
+    caplog.set_level(logging.ERROR)
+    mock_demo_notify.return_value = None
+    await setup_notify(hass)
+    await hass.async_block_till_done()
+    assert mock_demo_notify.called
+    assert "Failed to initialize notification service demo" in caplog.text
+
+
+async def test_discover_notify(hass, mock_demo_notify):
+    """Test discovery of notify demo platform."""
+    assert notify.DOMAIN not in hass.config.components
+    mock_demo_notify.return_value = None
+    discovery.load_platform(
+        hass, "notify", "demo", {"test_key": "test_val"}, {"notify": {}}
+    )
+    await hass.async_block_till_done()
+    assert notify.DOMAIN in hass.config.components
+    assert mock_demo_notify.called
+    assert mock_demo_notify.mock_calls[0][1] == (
+        hass,
+        {},
+        {"test_key": "test_val"},
+    )
+
+
+async def test_sending_none_message(hass, events):
+    """Test send with None as message."""
+    await setup_notify(hass)
+    with pytest.raises(vol.Invalid):
+        await hass.services.async_call(
+            notify.DOMAIN, notify.SERVICE_NOTIFY, {notify.ATTR_MESSAGE: None}
+        )
+    await hass.async_block_till_done()
+    assert len(events) == 0
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=6356951)